amethyst_renderer/src/visibility.rs

```diff
@@ -0,0 +1,109 @@
+use hibitset::BitSet;
+use specs::{Entities, Entity, Fetch, FetchMut, Join, ReadStorage, System};
+
+use cam::{ActiveCamera, Camera};
+use transparent::Transparent;
+
+/// Resource for controlling what entities should be rendered, and whether to draw them ordered or
+/// not, which is useful for transparent surfaces.
+#[derive(Default)]
+pub struct Visibility {
+    /// Visible entities that can be drawn in any order
+    pub visible_unordered: BitSet,
+    /// Visible entities that need to be drawn in the given order
+    pub visible_ordered: Vec<Entity>,
+}
+
+/// Determine what entities are visible to the camera, and which are not. Will also sort transparent
+/// entities back to front based on distance from camera.
+///
+/// Note that this should run after `GlobalTransform` has been updated for the current frame, and
+/// before rendering occurs.
+pub struct VisibilitySortingSystem {
+    centroids: Vec<Internals>,
+    transparent: Vec<Internals>,
+}
+
+#[derive(Clone)]
+struct Internals {
+    entity: Entity,
+    transparent: bool,
+    centroid: Point3<f32>,
+    camera_distance: f32,
+    from_camera: Vector3<f32>,
+}
+
+impl VisibilitySortingSystem {
+    /// Create new sorting system
+    pub fn new() -> Self {
+        VisibilitySortingSystem {
+            centroids: Vec::default(),
+            transparent: Vec::default(),
+        }
+    }
+}
+
+impl<'a> System<'a> for VisibilitySortingSystem {
+    type SystemData = (
+        Entities<'a>,
+        FetchMut<'a, Visibility>,
+        Option<Fetch<'a, ActiveCamera>>,
+        ReadStorage<'a, Camera>,
+        ReadStorage<'a, Transparent>,
+        ReadStorage<'a, GlobalTransform>,
+    );
+
+    fn run(
+        &mut self,
+        (entities, mut visibility, active, camera, transparent, global): Self::SystemData,
+    ) {
+        let origin = Point3::origin();
+
+        let camera: Option<&GlobalTransform> = active
+            .and_then(|a| global.get(a.entity))
+            .or_else(|| (&camera, &global).join().map(|cg| cg.1).next());
+        let camera_forward = camera
+            .map(|c| c.0.z.truncate())
+            .unwrap_or(Vector3::unit_z());
+        let camera_centroid = camera
+            .map(|g| g.0.transform_point(origin))
+            .unwrap_or(origin.clone());
+
+        self.centroids.clear();
+        self.centroids.extend(
+            (&*entities, &global)
+                .join()
+                .map(|(entity, global)| (entity, global.0.transform_point(origin)))
+                .map(|(entity, centroid)| Internals {
+                    entity,
+                    transparent: transparent.get(entity).is_some(),
+                    centroid,
+                    camera_distance: centroid.distance2(camera_centroid),
+                    from_camera: centroid - camera_centroid,
+                })
+                .filter(|c| c.from_camera.dot(camera_forward) > 0.), // filter entities behind the camera
```

Comment:
Doesn't take into account the camera FOV.
If its > 180, you can see behind the camera.
If its < 180, you might not be culling things that could be culled.

Reply:
This is known, to handle all those cases requires a much larger amount of math, which basically requires bounding volumes etc, which I don't want to add here. This is a system for fast prototyping, and will with 99% likelyhood not be used in larger games.

Reply:
Note that the `Visibility` structure can handle those since it has no concept of all that, it's up to the culling system to do all those things. Note that amethyst-rhusics is likely to have a better implementation for this, since there we can do real frustum culling based on the bounding volumes.